tests/gpflow/posteriors/test_posteriors.py

```diff
@@ -0,0 +1,445 @@
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from dataclasses import dataclass
+from inspect import isabstract
+from itertools import chain
+
+import numpy as np
+import pytest
+import tensorflow as tf
+
+import gpflow
+import gpflow.ci_utils
+from gpflow.models.util import inducingpoint_wrapper
+from gpflow.posterior import (
+    AbstractPosterior,
+    FallbackIndependentLatentPosterior,
+    FullyCorrelatedPosterior,
+    IndependentPosteriorMultiOutput,
+    IndependentPosteriorSingleOutput,
+    LinearCoregionalizationPosterior,
+)
+
+NUM_INDUCING_POINTS = 3
+
+
+@pytest.fixture(name="mean_function", params=[None, 1.0])
+def _mean_function_fixture(request):
+    mean_function_parameter = request.param
+
+    if mean_function_parameter is None:
+        return None
+    else:
+        return gpflow.mean_functions.Constant(mean_function_parameter)
+
+
+@pytest.fixture(name="whiten", params=[False, True])
+def _whiten_fixture(request):
+    return request.param
+
+
+@pytest.fixture(name="precompute", params=[False, True])
+def _precompute_fixture(request):
+    return request.param
+
+
+_independent_single_output = [IndependentPosteriorSingleOutput]
+_fully_correlated_multi_output = [FullyCorrelatedPosterior]
+_independent_multi_output = [IndependentPosteriorMultiOutput]
+_fallback_independent_multi_output = [FallbackIndependentLatentPosterior]
+_linear_coregionalization = [LinearCoregionalizationPosterior]
+
+
+def test_no_missing_kernels():
+    tested_kernels = set(
+        chain(
+            _independent_single_output,
+            _fully_correlated_multi_output,
+            _independent_multi_output,
+            _fallback_independent_multi_output,
+            _linear_coregionalization,
+        )
+    )
+
+    available_kernels = list(gpflow.ci_utils.subclasses(AbstractPosterior))
+    concrete_kernels = set([k for k in available_kernels if not isabstract(k)])
+
+    assert tested_kernels == concrete_kernels
+
+
+@dataclass(frozen=True)
+class LatentVariationalParameters:
+    qsqrt = (np.random.randn(NUM_INDUCING_POINTS, 1) ** 2) * 0.01
+    qmean = np.random.randn(NUM_INDUCING_POINTS, 1)
+
+
+def _assert_fused_predict_f_equals_precomputed_predict_f(
+    posterior, precompute, full_cov, full_output_cov, decimals=6
+):
+    Xnew = np.random.randn(20, 1)
+
+    fused_f_mean, fused_f_cov = posterior.fused_predict_f(
+        Xnew, full_cov=full_cov, full_output_cov=full_output_cov
+    )
+
+    if not precompute:
+        posterior.update_cache()
+
+    precomputed_f_mean, precomputed_f_cov = posterior.predict_f(
+        Xnew, full_cov=full_cov, full_output_cov=full_output_cov
+    )
+
+    np.testing.assert_array_almost_equal(fused_f_mean, precomputed_f_mean, decimal=decimals)
+    np.testing.assert_array_almost_equal(fused_f_cov, precomputed_f_cov, decimal=decimals)
```

Comment:
These are deprecated, see https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_almost_equal.html

Reply:
Good point. I have changed this function to use `assert_allclose`.